spec/plans/sync.fmf

```diff
@@ -0,0 +1,56 @@
+summary: Define synchronization method for each TMT step
+
+description: |
+    The main idea of :ref:`/spec/plans/provision/multihost`
+    tests is to run certain tasks in parallel (e.g., a RestAPI
+    requests from a client to a server). Therefore, these kinds
+    of tests may require test synchronization.
+    The synchronization method may be specified for each step
+    by providing one of defined options under ``step`` keyword.
+    This creates a synchronization barrier before each provided step.
+
+        start
+            Synchronize at the beginning of the ``step`` by adding
+            a synchronization barrier at the beginning of the step.
+            In case of multiple configurations, it also adds a barrier
+            at the beginning of the ``step``, even if the configuration
+            filters out every ``test``.
+        name
+            Synchronize execution of tests with the same name. Standalone
+            tests (without parallel test on a different guest) instantly
+            pass synchronization barrier. ``tmt`` is not responsible for
+            possible synchronization problems caused by duplicate test names
+            and similar mistakes.
+        index
+            Synchronize execution of all tests by adding a synchronization
+            barriers in front of each ``test`` execution.
+        none
+            **Default** value, which disables the synchronization for
+            given ``step``.
```

Comment:
In general, I lack more details about the properties of this "synchronization barrier", namely conditions required to match to pass the barrier. Ok, the general idea is clear, there would be some kind of synchronization points in tmt's workflow, but how do they behave, and what actors are involved?

```
        start
            Synchronize at the beginning of the ``step`` by adding
            a synchronization barrier at the beginning of the step.
            In case of multiple configurations, it also adds a barrier
            at the beginning of the ``step``, even if the configuration
            filters out every ``test``.
```

* The literal `step` in this paragraph seems misleading to me, I'd expect a mere "step" here rather than a keyword-looking `step`.
* Workflow would suspend before entering the step with `sync: step` - what I'm missing here: when would it resume? Is it waiting for some events? Or for other steps to catch up? If so, which events, which steps, the same "kind" of steps? All of them must reach the same barrier to proceed, or just one? Which one? If there are multiple step configurations defined (e.g. 5 `prepare` configs, 2 with `where: server`, 3 with `where: client`), what's the expected outcome? Shall `server` and `client` wait till they both reach `prepare` in general and move in their own speed, or would their configs run in lockstep? What happens to the third client's `prepare` config, will it wait for something, e.g. for the server's 2nd config to complete?

I'm afraid all this needs to be specified - it should be fairly doable to implement any set of rules, more or less complicated, but synchronization does have many aspects and corner cases, and the specification here seems like a tautology to me: "Synchronize at the beginning of the ``step`` by adding a synchronization barrier at the beginning of the step." - sure, but I didn't learn anything new after reading this sentence :)

WRT `name` and `test` - these seem like areas we're supposed to think about over Christmas :) To me, it's something that's probably orthogonal to step-level synchronization which could be specified properly first, before diving into test-level synchronization. "Run tests in lockstep" can then enable "run (at least) `execute` steps in lockstep", for example - step-level might be implemented once fully specified, as a building block to be used later by test-level synchronization. It seems less hairy, and our expectations seem to be more aligned than when it comes to test-level synchronization. The current specification draft leaves many questions unanswered, and that's just the step-level specification. My 2 cents would focus on step-level synchronization first, leaving test-level for later. I already have a rough draft of step-level synchronization, I'm missing the specification to follow :) Tests will be harder...